modules/fedimint-ln-client/src/recurring/mod.rs

```diff
@@ -0,0 +1,525 @@
+use fedimint_core::secp256k1::{Keypair, PublicKey};
+use fedimint_core::task::sleep;
+use fedimint_core::util::{BoxFuture, FmtCompact, SafeUrl};
+use fedimint_derive_secret::ChildId;
+use futures::StreamExt;
+use futures::future::select_all;
+use lightning_invoice::Bolt11Invoice;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use tokio::select;
+use tokio::sync::Notify;
+use tracing::{debug, trace};
+
+use crate::db::{RecurringPaymentCodeKey, RecurringPaymentCodeKeyPrefix};
+use crate::receive::LightningReceiveError;
+use crate::{
+    LightningClientModule, LightningClientStateMachines, LightningOperationMeta,
+    LightningOperationMetaVariant, LnReceiveState, tweak_user_key, tweak_user_secret_key,
+};
+
+impl LightningClientModule {
+    pub async fn register_recurring_payment_code(
+        &self,
+        protocol: RecurringPaymentProtocol,
+        recurringd_api: SafeUrl,
+        meta: &str,
+    ) -> Result<RecurringPaymentCodeEntry, RecurringdApiError> {
+        self.client_ctx
+            .module_db()
+            .autocommit(
+                |dbtx, _| {
+                    let recurringd_api_inner = recurringd_api.clone();
+                    let new_recurring_payment_code = self.new_recurring_payment_code.clone();
+                    Box::pin(async move {
+                        let next_idx = dbtx
+                            .find_by_prefix_sorted_descending(&RecurringPaymentCodeKeyPrefix)
+                            .await
+                            .map(|(k, _)| k.derivation_idx)
+                            .next()
+                            .await
+                            .map_or(0, |last_idx| last_idx + 1);
+
+                        let payment_code_root_key = self.get_payment_code_root_key(next_idx);
+
+                        let recurringd_client = RecurringdClient::new(recurringd_api_inner.clone());
+                        let register_response = recurringd_client
+                            .register_recurring_payment_code(
+                                self.client_ctx
+                                    .get_config()
+                                    .await
+                                    .global
+                                    .calculate_federation_id(),
+                                protocol,
+                                crate::recurring::PaymentCodeRootKey(
+                                    payment_code_root_key.public_key(),
+                                ),
+                                meta,
+                            )
+                            .await?;
+
+                        let payment_code_entry = RecurringPaymentCodeEntry {
+                            protocol,
+                            root_keypair: payment_code_root_key,
+                            code: register_response.recurring_payment_code,
+                            recurringd_api: recurringd_api_inner,
+                            last_derivation_index: 0,
```

Comment:
What does this do? At initial glance it looks like there is a derivation index in both the key and the value of this database entry

Reply:
If I'm reading correctly it looks like the derivation index in the key is for different payment codes, whereas the index in the value is for different invoices. Changing some naming might be helpful here